plugin/src/main/scala/chisel3/internal/plugin/IdentifierComponent.scala

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.internal.plugin
+
+import scala.collection.mutable
+import scala.reflect.internal.Flags
+import scala.tools.nsc
+import scala.tools.nsc.{Global, Phase}
+import scala.tools.nsc.plugins.PluginComponent
+import scala.tools.nsc.transform.TypingTransformers
+
+// The component of the chisel plugin. Not sure exactly what the difference is between
+//   a Plugin and a PluginComponent.
+class IdentifierComponent(val global: Global, arguments: ChiselPluginArguments)
+    extends PluginComponent
+    with TypingTransformers {
+  import global._
+  val runsAfter: List[String] = "typer" :: Nil
+  val phaseName: String = "identifiercomponent"
+  def newPhase(_prev: Phase): ChiselComponentPhase = new ChiselComponentPhase(_prev)
+  class ChiselComponentPhase(prev: Phase) extends StdPhase(prev) {
+    override def name: String = phaseName
+    def apply(unit: CompilationUnit): Unit = {
+      if (ChiselPlugin.runComponent(global, arguments)(unit)) {
+        unit.body = new MyTypingTransformer(unit).transform(unit.body)
+      }
+    }
+  }
+
+  private class MyTypingTransformer(unit: CompilationUnit) extends TypingTransformer(unit) {
+
+    def inferType(t: Tree): Type = localTyper.typed(t, nsc.Mode.TYPEmode).tpe
+
+    val baseModuleTpe: Type = inferType(tq"chisel3.experimental.BaseModule")
+    val stringTpe:     Type = inferType(tq"String")
+
+    private def stringFromTypeName(name: TypeName): String =
+      name.toString.trim() // Remove trailing space (Scalac implementation detail)
+
+    // Not cached because it should only be run once per class (thus once per Type)
+    def isAModule(sym: Symbol): Boolean = { sym.tpe <:< baseModuleTpe }
+    def isExactBaseModule(sym: Symbol): Boolean = { sym.tpe =:= baseModuleTpe }
+
+    def isNullaryMethodNamed(name: String, defdef: DefDef): Boolean =
+      defdef.name.decodedName.toString == name && defdef.tparams.isEmpty && defdef.vparamss.isEmpty
+
+    def getConstructorAndParams(body: List[Tree]): (Option[DefDef], Seq[Symbol]) = {
+      val paramAccessors = mutable.ListBuffer[Symbol]()
+      var primaryConstructor: Option[DefDef] = None
+      body.foreach {
+        case acc: ValDef if acc.symbol.isParamAccessor && !acc.mods.hasFlag(Flag.BYNAMEPARAM) =>
+          paramAccessors += acc.symbol
+        case con: DefDef if con.symbol.isPrimaryConstructor =>
+          primaryConstructor = Some(con)
+        case d: DefDef if isNullaryMethodNamed("_moduleDefinitionIdentifierProposal", d) =>
+          val msg = "Users cannot override _moduleDefinitionIdentifierProposal. Let the compiler plugin generate it."
+          global.reporter.error(d.pos, msg)
+        case _ =>
+      }
+      (primaryConstructor, paramAccessors.toList)
+    }
+
+    def generateIdentifierMethod(module: ClassDef, thiz: global.This, baseClass: global.Symbol): Tree = {
```

Comment:
I don't love how much code we've duplicated between the other components and this one, maybe we can make a utils object?

Reply:
One issue is all the types are imported from `import global._`, so its hard to actually do this well.

Reply:
In the utils, you can take `global` as an argument and use it to derive types, see https://github.com/chipsalliance/chisel/blob/3a59415380a75085286e5cbfd006323ea04ea02d/core/src/main/scala/chisel3/Printf.scala#L39

Reply:
Sounds good.